helix-core/src/main/java/org/apache/helix/task/TaskStateModelFactory.java

```diff
@@ -102,4 +106,73 @@ public boolean isShutdown() {
   public boolean isTerminated() {
     return _taskExecutor.isTerminated();
   }
+
+  /*
+   * Get target thread pool size from InstanceConfig first; if that fails, get it from
+   * ClusterConfig; if that fails, fall back to the default value.
+   */
+  private int getTaskThreadPoolSize() {
+    ConfigAccessor configAccessor = _manager.getConfigAccessor();
+    // Check instance config first for thread pool size
+    InstanceConfig instanceConfig =
+        configAccessor.getInstanceConfig(_manager.getClusterName(), _manager.getInstanceName());
+    if (instanceConfig != null) {
+      int targetTaskThreadPoolSize = instanceConfig.getTargetTaskThreadPoolSize();
+      if (verifyTargetThreadPoolSize(targetTaskThreadPoolSize)) {
+        return targetTaskThreadPoolSize;
+      }
+    }
+
+    // Fallback to cluster config since instance config doesn't provide the value
+    ClusterConfig clusterConfig = configAccessor.getClusterConfig(_manager.getClusterName());
+    if (clusterConfig != null) {
+      int targetTaskThreadPoolSize = clusterConfig.getDefaultTargetTaskThreadPoolSize();
+      if (verifyTargetThreadPoolSize(targetTaskThreadPoolSize)) {
+        return targetTaskThreadPoolSize;
+      }
+    }
+
+    return TaskConstants.DEFAULT_TASK_THREAD_POOL_SIZE;
+  }
+
+  /*
+   * Checks against the default values of -1 when pool sizes are not defined; we don't want -1's
+   */
+  private static boolean verifyTargetThreadPoolSize(int targetTaskThreadPoolSize) {
+    return targetTaskThreadPoolSize > 0;
+  }
+
+  /*
+   * Update LiveInstance with the current used thread pool size
+   */
+  private void updateLiveInstanceWithThreadPoolSize(int taskThreadPoolSize) {
```

Comment:
Instead of updating after the live node created, we shall ensure it is included when the node is created the first time. Or we may have some serious issue that the controller violates the capacity limitation.

Reply:
This is a good point. Let me think about this.